cgi-bin/upperCase.sh

```diff
@@ -0,0 +1,8 @@
+#!/bin/bash
+
+# Set the content type for the response (CGI header)
+echo "Content-Type: text/plain"
+echo ""
```

Comment:
Yes, a CGI script should have a Content-Type header in its response telling the client what kind of content is being returned.

Good point about the CRLF. Changed to printf which should be more portable.

Reply:
But shouldn't the content-type header be part of the response header?
If the content-type header is returned in the body the client won't know, until the body is read. But he doesn't know how to read it, since he doesn't know the type.

Put in another way: shouldn't we tell the client in the **response header** what content to expect? 

Right now the content-type header is added depending on the MIME type / file extension. 
So we would have one content-type in the response header and another one in the response body.

Reply:
It will be part of the response header, as the last line of it to be precise.

This is reflected in the appendHeaders() method of the ResponseBuilder:
```
if (!request.hasCGI && (m_responseBody.find("Content-Type: ") == std::string::npos))
		m_responseStream << "Content-Type: " << getMIMEType(webutils::getFileExtension(request.targetResource))
						 << "\r\n";
```
The CGI script simply creates this one header field and the rest of the headers will be created by the server.

Reply:
The response header is delimited by 
\r\n (last header) 
\r\n (emtpy line)

Right now we just check if we find the header in the body. If we do, we don't add it. But does this affect the other headers? I think the client receives:
```
<status line>\r\n
<server created header1>\r\n
<server created headerN>\r\n
\r\n // "official" end of response header
<content-type "header">\r\n
\r\n
<rest of body body>
```

Since the header portion is delimited by \r\n how would the client know, that there is an additional header to parse?

Reply:
The idea is to have it like that:
```
<status line>\r\n
<server created header1>\r\n
<server created headerN>\r\n
<content-type "header">\r\n
\r\n
<rest of body body>
```

To achieve that, it shall also be verified if there is a Content-Type header in the returned "body":

```
// Delimiter
if (!request.hasCGI && (m_responseBody.find("Content-Type: ") == std::string::npos))
	m_responseStream << "\r\n";
```

Reply:
I suggest to extract the cgi-header logic:
```
// client requested CGI and it was successful
if (isCGI && response.status == StatusOK)
  appendHeadersCGI(...)
else
  appendHeaders(...)
```

Right now we have to check the whole body of a response for the header, irregardless of CGI (e.g. we have to check a .png binary file for a content type header).


Somewhere before we could make sure the content-type header is present > seems to be required? 
RFC 3875 Section 6.3 [Response Header Fields ](https://datatracker.ietf.org/doc/html/rfc3875#section-6.3) calls em _CGI-field_ which is equivalent to Content-Type | Location | Status. The other stuff seems like we could get away with not implementing.

I think ResponseBodyHandler would be a good fit
```
[...]
} else if (m_request.hasCGI) {
  if (m_responseBody.find("Content-Type: ") == std::string::npos)) {
// we didn't find header, so it errors  
  request.status = StatusBadRequest;
// since error we need to generate an error body
  handleErrorBody();
} else if (m_request.hasAutoindex) {
[...]
```

Sidenote: here we have again the case sensitive search problem tracked in #77

Reply:
It introduces another function which is very similar to appendHeaders(), but if you prefer it this way, I don't mind. Implemented.

Reply:
You can also do the check inside of `appendHeaders()` if you prefer. But I think only checking the body for stuff if we can expect to find something in there is a reasonable change and nearly free performance boost

Reply:
No it is fine, we can leave it as is;